mps_youtube/players/mpv.py

```diff
@@ -0,0 +1,298 @@
+
+            if g.mprisctl:
+                g.mprisctl.send(('socket', self.sockpath))
+        else:
+            if g.mprisctl:
+                self.fifopath = tempfile.mktemp('.fifo', 'mpsyt-mpv')
+                os.mkfifo(self.fifopath)
+                cmd.append('--input-file=' + self.fifopath)
+                g.mprisctl.send(('mpv-fifo', self.fifopath))
+
+            self.p = subprocess.Popen(cmd, shell=False, stderr=subprocess.PIPE,
+                                      bufsize=1)
+
+        self._player_status(self.songdata + "; ", self.song.length)
+        returncode = self.p.wait()
+
+        if returncode == 42:
+            self.previous()
+
+        elif returncode == 43:
+            self.stop()
+
+        else:
+            self.next()
+
+    def _player_status(self, prefix, songlength=0):
+        """ Capture time progress from player output. Write status line. """
+        # pylint: disable=R0914, R0912
+        re_player = re.compile(r".{,15}AV?:\s*(\d\d):(\d\d):(\d\d)")
+        re_volume = re.compile(r"Volume:\s*(?P<volume>\d+)\s*%")
+        last_displayed_line = None
+        buff = ''
+        volume_level = None
+        last_pos = None
+
+        if self.sockpath:
+            s = socket.socket(socket.AF_UNIX)
+
+            tries = 0
+            while tries < 10 and self.p.poll() is None:
+                time.sleep(.5)
+                try:
+                    s.connect(self.sockpath)
+                    break
+                except socket.error:
+                    pass
+                tries += 1
+            else:
+                return
+
+            try:
+                observe_full = False
+                cmd = {"command": ["observe_property", 1, "time-pos"]}
+                s.send(json.dumps(cmd).encode() + b'\n')
+                volume_level = elapsed_s = None
+
+                for line in s.makefile():
+                    resp = json.loads(line)
+
+                    # deals with bug in mpv 0.7 - 0.7.3
+                    if resp.get('event') == 'property-change' and not observe_full:
+                        cmd = {"command": ["observe_property", 2, "volume"]}
+                        s.send(json.dumps(cmd).encode() + b'\n')
+                        observe_full = True
+
+                    if resp.get('event') == 'property-change' and resp['id'] == 1:
+                        if resp['data'] is not None:
+                            elapsed_s = int(resp['data'])
+
+                    elif resp.get('event') == 'property-change' and resp['id'] == 2:
+                        volume_level = int(resp['data'])
+
+                    if(volume_level and volume_level != g.volume):
+                        g.volume = volume_level
+                    if elapsed_s:
+                        line = self._make_status_line(elapsed_s, prefix, songlength,
+                                                      volume=volume_level)
+
+                        if line != last_displayed_line:
```

Comment:
It would be good if the  `_make_status_line()`, `if line != last_displayed_line`, `screen.writestatus()` logic was all handled in `self._make_status_line()`. Or is there a better design that removes this redundancy?

Reply:
I warped the `_make_status_line()` in `make _status_line()`. Probably not the best fix, but it works.